cpp/src/experimental/graph.cu

```diff
@@ -0,0 +1,543 @@
+}
+
+template <typename vertex_t, typename DegreeIterator, typename ThresholdIterator>
+std::vector<vertex_t> segment_degree_sorted_vertex_partition(raft::handle_t const &handle,
+                                                             DegreeIterator degree_first,
+                                                             DegreeIterator degree_last,
+                                                             ThresholdIterator threshold_first,
+                                                             ThresholdIterator threshold_last)
+{
+  auto num_elements = thrust::distance(degree_first, degree_last);
+  auto num_segments = thrust::distance(threshold_first, threshold_last) + 1;
+
+  std::vector<vertex_t> h_segment_offsets(num_segments + 1);
+  h_segment_offsets[0]     = 0;
+  h_segment_offsets.back() = num_elements;
+
+  rmm::device_uvector<vertex_t> d_segment_offsets(num_segments - 1, handle.get_stream());
+
+  thrust::upper_bound(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                      degree_first,
+                      degree_last,
+                      threshold_first,
+                      threshold_last,
+                      d_segment_offsets.begin());
+
+  raft::update_host(h_segment_offsets.begin() + 1,
+                    d_segment_offsets.begin(),
+                    d_segment_offsets.size(),
+                    handle.get_stream());
+
+  return h_segment_offsets;
+}
+
+}  // namespace
+
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu>
+graph_t<vertex_t, edge_t, weight_t, store_transposed, multi_gpu, std::enable_if_t<multi_gpu>>::
+  graph_t(raft::handle_t const &handle,
+          std::vector<edgelist_t<vertex_t, edge_t, weight_t>> const &edgelists,
+          partition_t<vertex_t> const &partition,
+          vertex_t number_of_vertices,
+          edge_t number_of_edges,
+          bool is_symmetric,
+          bool is_multigraph,
+          bool is_weighted,
+          bool sorted_by_global_degree_within_vertex_partition,
+          bool do_expensive_check)
+  : detail::graph_base_t<vertex_t, edge_t, weight_t>(
+      handle, number_of_vertices, number_of_edges, is_symmetric, is_multigraph, is_weighted),
+    partition_(partition)
+{
+  // cheap error checks
+
+  auto &comm_p         = this->get_handle_ptr()->get_comms();
+  auto comm_p_size     = comm_p.get_size();
+  auto &comm_p_row     = this->get_handle_ptr()->get_subcomm(comm_p_row_key);
+  auto comm_p_row_rank = comm_p_row.get_rank();
+  auto comm_p_row_size = comm_p_row.get_size();
+  auto &comm_p_col     = this->get_handle_ptr()->get_subcomm(comm_p_col_key);
+  auto comm_p_col_rank = comm_p_col.get_rank();
+  auto comm_p_col_size = comm_p_col.get_size();
+  auto default_stream  = this->get_handle_ptr()->get_stream();
+
+  for (size_t i = 0; i < edgelists.size(); ++i) {
```

Comment:
Perhaps replace the `for` loop with a `find_if()` on the `edgelist` container?

Reply:
replaced the for loop with std::any_of.